codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/protocol/ProtocolTestGenerator.kt

```diff
@@ -166,33 +200,56 @@ class ProtocolTestGenerator(
         rustTemplate(
             """
-            let builder = #{config}::Config::builder().with_test_defaults().endpoint_resolver("https://example.com");
+            let (conn, request_receiver) = #{capture_request}(None);
+            let config_builder = #{config}::Config::builder().with_test_defaults().endpoint_resolver("https://example.com");
             #{customParams}
-            let config = builder.build();
 
             """,
+            "capture_request" to CargoDependency.smithyClient(codegenContext.runtimeConfig)
+                .toDevDependency()
+                .withFeature("test-util")
+                .toType()
+                .resolve("test_connection::capture_request"),
             "config" to ClientRustModule.Config,
             "customParams" to customParams,
         )
-        writeInline("let input =")
-        instantiator.render(this, inputShape, httpRequestTestCase.params)
+        renderClientCreation(this, ClientCreationParams(codegenContext, "conn", "config_builder", "client"))
+
+        writeInline("let result = ")
+        instantiator.renderFluentCall(this, "client", operationShape, inputShape, httpRequestTestCase.params)
+        rust(""".send().await;""")
+        // Response parsing will always fail since we feed it an empty response body, so we don't care
+        // if it fails, but it is helpful to print what that failure was for debugging
+        rust("let _ = dbg!(result);")
+        rust("""let http_request = request_receiver.expect_request();""")
 
-        rust(""".make_operation(&config).await.expect("operation failed to build");""")
-        rust("let (http_request, parts) = input.into_request_response().0.into_parts();")
         with(httpRequestTestCase) {
             host.orNull()?.also { host ->
                 val withScheme = "http://$host"
+                when (val bindings = EndpointPrefixGenerator.endpointTraitBindings(codegenContext, operationShape)) {
+                    null -> rust("let endpoint_prefix = None;")
+                    else -> {
+                        withBlock("let input = ", ";") {
+                            instantiator.render(this@renderHttpRequestTestCase, inputShape, httpRequestTestCase.params)
+                        }
+                        withBlock("let endpoint_prefix = Some({", "}.unwrap());") {
+                            bindings.render(this, "input", codegenContext.smithyRuntimeMode, generateValidation = false)
+                        }
+                    }
+                }
                 rustTemplate(
                     """
                     let mut http_request = http_request;
                     let ep = #{SmithyHttp}::endpoint::Endpoint::mutable(${withScheme.dq()}).expect("valid endpoint");
-                    ep.set_endpoint(http_request.uri_mut(), parts.acquire().get()).expect("valid endpoint");
+                    ep.set_endpoint(http_request.uri_mut(), endpoint_prefix.as_ref()).expect("valid endpoint");
```

Comment:
do we still need to do it this way now that we're capturing the entire request?

Reply:
It's needed for cases like this where the test supplies a custom host with a path in it:
https://github.com/awslabs/smithy/blob/be68f3bbdfe5bf50a104b387094d40c8069f16b1/smithy-aws-protocol-tests/model/restJson1/endpoint-paths.smithy#L19